doc/sphinxext/missing_references.py

```diff
@@ -0,0 +1,252 @@
+        line = str(line)
+    else:
+        line = ""
+
+    return "%s:%s" % (path, line)
+
+
+def save_missing_references_handler(app, exc):
+    """
+    At the end of the sphinx build, either save the missing references to a
+    JSON file. Also ensure that all lines of the existing JSON file are still
+    necessary.
+    """
+    if not app.config.missing_references_enabled:
+        # no-op when we are disabled.
+        return
+
+    json_path = os.path.join(app.confdir,
+                             app.config.missing_references_filename)
+
+    references_warnings = getattr(app.env, 'missing_reference_warnings', {})
+
+    # This is a dictionary of {(dtype,target): locations}
+    references_ignored = getattr(app.env,
+        'missing_references_ignored_references', {})
+    references_events = getattr(app.env, 'missing_reference_events', {})
+
+    # Warn about any reference which is no longer missing.
+    for (dtype, target), locations in references_ignored.items():
+        missing_reference_locations = references_events.get(
+                (dtype, target), [])
+
+        # For each ignored reference location, ensure a missing reference was
+        # observed. If it wasn't observed, issue a warning.
+        for ignored_refernece_location in locations:
+            if ignored_refernece_location not in missing_reference_locations:
+                msg = (f"Reference {dtype} {target} for "
+                       f"{ignored_refernece_location} can be removed"
+                       f" from {app.config.missing_references_filename}."
+                        "It is no longer a missing reference in the docs.")
+                logger.warning(msg,
+                    location=ignored_refernece_location,
+                    type='ref',
+                    subtype=dtype)
+
+    if app.config.missing_references_write_json:
+        _write_missing_references_json(references_warnings, json_path)
+
+
+def _write_missing_references_json(records, json_path):
+    """
+    Convert ignored references to a format which we can write as JSON
+
+    Convert from ``{(dtype, target): locaitons}`` to
+    ``{dtype: {target: locations}}`` since JSON can't serialize tuples.
+    """
+    transformed_records = defaultdict(dict)
+
+    for (dtype, target), paths in records.items():
+        paths = list(paths)
+        paths.sort()
+        transformed_records[dtype][target] = paths
+
+    with open(json_path, "w") as stream:
+        json.dump(transformed_records, stream, indent=2)
+
+
+def _read_missing_references_json(json_path):
+    """
+    Convert from the JSON file to the form used internally by this
+    extension.
+
+    The JSON file is stored as ``{dtype: {target: [locations,]}}`` since JSON
+    can't store dictionary keys which are tuples. We convert this back to
+    ``{(dtype,target):[locations]}`` for internal use.
+
+    """
+    with open(json_path, "r") as stream:
+        data = json.load(stream)
+
+    ignored_references = {}
```

Comment:
would write this as a (nested) dict comprehension but heh

Reply:
I think I tried that, and didn't like the line folding I had to do to get under 80 characters.